### DISCUSSION TITLE
Why doesn't Fastify allow plugins to return a value as the promised value of fastify.register?

### DISCUSSION THREAD

**Discussion by Worldwidebrine** [fastify/fastify#4686] [Q&A]
`FastifyPluginCallback` and `FastifyPluginAsync` return `void`.
`fastify.register(...)` returns `FastifyInstance<...> & PromiseLike<undefined>`, so `await fastify.register(...)` is always `undefined`.

Why doesn't Fastify allow plugins to return a value and let it be the promised value of `fastify.register`?

```js
const returnedValue = fastify.register((fastify, opts, done) => {
  setTimeout(() => done(undefined, "test"), 10);
  // or async return "test";
});
const promisedValue = await returnedValue;
console.log(returnedValue, promisedValue);
```

According to .d.ts files, the result should be `[object Object] undefined`, but actually `[object Object] [object Object]` both of them are objects.
Why shouldn't it be `[object Object] test`?

**Comment by metcoder95** [comment]
Indeed, it seems to be a mismatch between the Types and what Fastify actually does.
The promise resolves the `fastify` instance after fulfillment.
Though, not so sure why the Types do not match.

In any case, can you elaborate what is the value you expect by allowing plugins to return a specific value after the plugin is registered? Just want to understand what are you trying to achieve 🙂

**Reply by Worldwidebrine** [reply]
This was just the question that occurred to me when I found the mismatch. I was thinking about something that becomes available after the plugin is fully loaded but I don't need anything specifically.

**Reply by metcoder95** [reply]
Got it 👍 
I'm not so sure about the intention the type mismatch was unintended or not. But feel free to open an issue to kick off a chat out of it
